Add a trivial ip address based acl

abogani opened the pull request.

abogani committed:
Add ip based rules acl

abogani commented on main.cpp:
```diff
@@ -100,6 +102,21 @@ int main(int argc,char *argv[])
+	int found_cfg = 0;
+	int k = 0;
+	while(k<argc && !found_cfg) {
+	  found_cfg = (strcasecmp(argv[k],"--cfgFile")==0);
```
Perhaps should we rename the option --cfgFile (which sounds a little generic) with (i.e.) --aclFile (which sounds a more bit  specific)?

Ingvord replied:
According to POSIX it should be `--acl-file`

abogani replied:
I thought that be consistent was the right approach. Please let me what is for you the best name for that option!

bourtemb commented:
I think we should keep the old behaviour (authorizing clients from everywhere) by default if the Database server is not started with this new option.
I might be wrong, but it looks like the proposed patch as it is now is imposing the usage of this configuration file and will refuse any access to the Tango database to any client if this option was not used.

abogani commented:
Good catch! You are right it was my fault. I'll change the patchset to fix that (wrong) behaviour.

abogani commented:
List of changes:
*) Follow (odd) code indentation 
*) Follow option names convention so -aclFile like the already present -maxFile and -poolSize
*) Reduce overhead to check a boolean for users who don't use that feature

bourtemb commented:
Thanks Alessio for the changes.
To be clean, I think is_acl_enabled variable should be initialized to false at some point.

abogani commented:
Global variables are zero-initialized by default.

bourtemb commented:
Right you are! In this case it should be fine.

abogani commented:
Please wait before merging: Graziano have spotted a bug in the code!

abogani commented:
Now the code should be in good state. What do you think about?

By the way preparing this patch I just noticed that inet_addr() is used in Tango's source code which AFAIK doesn't support IPv6. Is Tango supposed to work properly with IPv6?

bourtemb commented:
I can't see any new commits since you commented the last time... Did you push your commit to your master branch on your fork (the one used when you created this Pull Request)?

You are right. Tango is currently not supposed to work properly with IPv6. IPv6 is currently not supported.

There are some tricks to make device servers run on IPv6 machines:
http://tango-controls.readthedocs.io/en/latest/tutorials-and-howtos/how-tos/how-to-run-device-servers-on-XP-with-IPv6.html

abogani commented:
Yes I did.

bourtemb commented:
OK. I was confused because the date of the commit is 30 Aug. But you pushed it yesterday indeed.
I was also confused because there is only one commit in this PR. I was expecting to get another commit with your latest changes. Then it would have been easier to review the last changes.
What did you change compared to the first version?

abogani commented:
You are right an additional commit would have been easier your review task. Sorry for missing to do it but I didn't known that GitHub makes possible to merge PR squashing its commits (Graziano has just told me right now).

Latest changes are:
*) UNIX domain sockets are always allowed (i.e. giop:unix:)
*) Configuration file could contains comments (#)
*) Addresses provided by configuration file are validated and assured to be in the right form converting into binary first (inet_pton()) and then (again) to text (inet_ntop())

By the way, the binary name was "DataBaseds" in old Tango versions and now it is "Databaseds", Is it desired?

Thanks!

bourtemb commented:
Thank you very much for the description of the changes...
You are right. The binary name used to be DataBaseds and was renamed Databaseds during the CMakenization process.
This can create some issues with the startup scripts which are provided with the TANGO distribution.
Of course, these scripts could be updated when preparing the next Tango distribution...
The question is: do we really need to update the name of the binary?
@Ingvord, do you have any strong opinion about it?
It would be easy to revert the name to the old one of course.

abogani commented:
Hi @bourtemb and @Ingvord,

Are there any news?

Thanks in advance!

Ingvord commented on main.cpp:
```diff
@@ -78,6 +174,30 @@ int main(int argc,char *argv[])
+	// Browse argv and try to find the --aclFile option
+	int found_acl = 0;
+	int k = 0;
+	while(k<argc && !found_acl) {
```
Should be **for**, not **while**

abogani replied:
Our first and most important rule is consistency: if there isn't a very good reason to change keep the things how we find. And this is what I have done with this patch. In other words in our opinion or we accept this form (we also follow already present indentation!) or we change all code which have already this form. This applies also on all other your comments: -maxFile and -poolSize for argument names, for use of the while **loop** instead of the **for** one, for extraction a method from a common few lines of code and for the name of the binary.

bourtemb replied:
I tend to agree with Alessio here for this PR (consistency is important) but we should keep in mind that this code could be improved indeed as suggested by Igor.
What I would suggest is to keep the code consistent for the next TANGO 9 releases. So I would even rename the executable to the name it was in Tango 9.2.5a official release. I would not rename an executable in a minor release.
For TANGO 10, we are more free and could indeed refactor the code, options names and executable name to follow the most followed standards.
> The nice thing about standards is that there are so many of them to choose from - Andrew S. Tanenbaum

Ingvord commented on main.cpp:
```diff
@@ -78,6 +174,30 @@ int main(int argc,char *argv[])
+	int found_acl = 0;
+	int k = 0;
+	while(k<argc && !found_acl) {
+	  found_acl = (strcasecmp(argv[k],"-aclFile")==0);
```
`-aclFile` is not POSIX argument, must be `--acl-file`

Ingvord commented on main.cpp:
```diff
@@ -78,6 +174,30 @@ int main(int argc,char *argv[])
+	  found_acl = (strcasecmp(argv[k],"-aclFile")==0);
+	  if(!found_acl) k++;
+	}
+	if( found_acl ) {
```
Should be extracted into a method

Ingvord commented on main.cpp:
```diff
@@ -54,11 +54,107 @@ static const char *RcsId = "$Id$";
+bool is_acl_enabled;
+std::set<std::string> ip_addresses_allowed;
+
+CORBA::Boolean retrieve_client_address(omni::omniInterceptors::serverReceiveRequest_T::info_T& info)
```
I think something like this is already implemented in the kernel. I guess that can be used in allow_command (or is_command_allowed). This will guarantee that every command is validated against the policy without inserting `check_ip` everywhere. Can be done in a separate PR though.

abogani replied:
We would prefer avoid that approach because we don't want check all Tangodatabaseds commands indiscriminately but only the ones which make changes to the RDBMS. In other words readings are always allowed plus a couple of commands that if not allowed make Tango stops to work.

bourtemb replied:
I don't know whether this is what Igor was suggesting but indeed the best location for the check_ip() calls would be in DataBaseStateMachine.cpp file in the is_xxx_allowed() methods (e.g.: DataBase::is_DbPutAttributeAlias_allowed ).
This would be cleaner and indeed could be done in separate PR.

gscalam replied:
Hi @bourtemb ,
if you mean to move the check in the is_xxx_alowed method and return just true/false we would loose some information we put in the exception description (the ip of the caller that is not allowed) that could be useful.
I am not sure of the behavior if we throw an exception in the is_xxx_allowed method.

bourtemb replied:
I agree with you on the fact we should not lose the information coming from the exception.
I was not considering returning false but throwing the exception in the is_xxx_allowed method.
I tested on a standard server and the exception is correctly propagated to the client.

Ingvord commented:
@bourtemb strictly speaking Unix naming convention prefers `-` or `_` to separate logical parts of the file name. So it should be _database-ds_ or _database_ds_ on unix platforms. Do not know about Windows though. We should ask windows expert...

abogani committed:
Move checking to DataBaseStateMachine.cpp

abogani commented:
I changed what requested from @Ingvord and @bourtemb about moving checking to DataBaseStateMachine inside _allowed methods.

If you merge please select squash and merge.

abogani committed:
Use an unordered_set instead of a set

abogani commented:
@bourtemb During the last meeting I erroneously swapped set with unordered_set (please let Goets knows about it). I'm still convinced that set is better (after all dataset si small and ip address could be very similar fooling the hashing algorithm). In any case if you plan to test this feature please execute it with and without following commit (which uses an unordered_set instead of a set). Sorry for mistake.

bourtemb commented:
Hi Alessio. Thanks for the clarification. I just tag @andygotz and I encourage him to have a look at your comment related to unordered_set if he didn't already read it.
It looks like you didn't have a chance to do any performance tests with unordered_set.
I think you were saying during the teleconf that you are using the version using sets since several months now and you didn't get any trouble... Just out of curiosity, how big is your set in your specific use case?

andygotz commented:
I understand this pull request addresses a real need for Elettra and any site which needs to protect changes to the Tango database. After discussion this need cannot be addressed by the TAC filtering mechanism because TAC is implemented in the client and a client can be modified e.g. setting an evironment variable to ignore the TAC. The code is simple but inserting a call into the commands which need to be protected looks a bit like a hack. It could have been done in a single place. The advantage of the way it has been done is it will slow down client access a bit only for the commands with a call to check_ip(). 

In principle we can accept the PR but before doing so we need comments in the code - none of the new methods have a single comment! We also need a small document explaining how to use it and giving examples. The document should be written in sphinx and added to the Tango documentation.

I worry that for sites with many hosts which need to access the database the list of IPs addresses to be specified is very long and could take a long time to check. But maybe I have misunderstood something.

The way it is implemented by checking only the ip address means that as soon as a device server is on a host all clients for that host are considered to be trusted. This makes it easy to get aroudn this check - the client simply runs on a host where a device server is running. This is why the TAC checks the user and group. 

How docker proof is this check? Can the ip address be spoofed using a docker container?

It might be worth considering modifying the TAC to implement the same feature in a safer way.

We wait for your comments to the code before merging.

gscalam commented:
Hi @bourtemb,
we have deployed it just on some beamlines, so we have around 10 addresses authorized R/W in the configuration file.
In any case in FERMI, our biggest use case at the moment, we will have around 170 addresses to authorize with R/W access.

abogani commented:
Hi @andygotz,

Could you provide me a example of a well-documented thing, please?

The list of the IP is checked only if user enable that feature. In any case our hope is that list of those IP would be very short. As Graziano already replied this is our case so it is true the feature slow down things but only a very little.

Our production systems are protected by password or by SSH key so there isn't a way for normal user to launch a Tango client on the same machine on which run a Tango server. 

A better solution (i.e. check the IP address only during initial connection) requires to make CORBA aware of concepts of the Tango Database but this would be very difficult to implement.

abogani committed:
Patch revisited:

*) Add namespace std:: prefix
*) Use OmniThread library instead of the pthreads one
*) Remove use of find_packages(Threads) into CMakeLists.txt
   to find pthreads library
*) Theoretically support any omniORB's transport layer (giop:tcp, giop:ssl,
   giop:fd, giop:unix, giop:http) at the moment tested against giop:tcp only
*) Rename funcions in accordance of the previous point
*) Make code more concise, shorter and add few comments
